Set up integration test framework

nkzarrabi opened the pull request:
Is this along the lines of what you were thinking of? Any issues with how I'm accessing the HTML or model classes?

nkzarrabi commented:
@kerkeslager

kerkeslager commented:
Yeah, I haven't actually run it, but this looks pretty good. A few suggestions:

We probably want to run the tests on both Chrome and Firefox, but we want to write them once. To do that, we can do this:

```
# don't inherit from TestCase
class IntegrationTests(object):
    def test_that_we_only_have_to_write_once(self):
        ...

# inherit from StaticLiveServerTestCase here so the tests get picked up by the test runner
# inherit from IntegrationTests so that it has all our tests
class FirefoxIntegrationTests(StaticLiveServerTestCase, IntegrationTests):
    def setUpClass(cls):
        # firefox specific setup here
    ...

class ChromeIntegrationTests(StaticLiveServerTestCase, IntegrationTests):
    def setUpClass(cls):
        # chrome specific setup here
```

kerkeslager commented:
We can also probably ignore `*.log` in the `.gitignore`.

nkzarrabi reviewed (commented):
In support of @kerkeslager revision, wrote one test as IntegrationTest, ex/(ab?)-stracting browser specific code from setUpClass

kerkeslager commented on core/test_integration.py:
```diff
@@ -0,0 +1,72 @@
+        # Wait until the home page is loaded
+        
+        
+        self.wait.until(EC.url_to_be(self.home_url))
```
Does this have some sort of timeout? If the page never loads or if the wrong URL loads, we want the test to fail, rather than hang forever and stop running all the tests.

kerkeslager replied:
I'm not sure, haven't used Selenium in a while to remember, but I'll trust your reading of the docs. Seems good to me.

kerkeslager commented on core/test_integration.py:
```diff
@@ -0,0 +1,72 @@
+    def setUp(self):
+        # Setup code to run before each test, e.g., creating test data
+        User = get_user_model()
+        self.user = User.objects.create_user(username='testuser', password='testpassword')
```
I might move the creation of the user into the actual test, since we'll probably want to test the signup form in this same class.

kerkeslager commented on core/test_integration.py:
```diff
@@ -0,0 +1,72 @@
+        User = get_user_model()
+        self.user = User.objects.create_user(username='testuser', password='testpassword')
+        self.login_url = self.live_server_url + reverse('login')  # 'login' is the name of your login URL
+        self.home_url = self.live_server_url + reverse('index')  # 'home' is the name of your post-login redirect URL
```
I would move these into the test as well--we can factor out any duplication in the tests after we have a few tests, but right now we only have one test so we don't know what the duplication will be.

kerkeslager commented on core/test_integration.py:
```diff
@@ -0,0 +1,72 @@
+    @classmethod
+    def setUpClass(cls):
+        # Setup code to run once before all tests, e.g., starting a browser
+        cls.browser = webdriver.Firefox(service=Service(GeckoDriverManager().install()))
```
I'm a little worried that setting up the browser at the class level might mean that the browser state is saved between tests, meaning one test could affect other tests (i.e. if one tests that log in works, the user will be logged in, which might cause a second test of logins to see that user still logged in, even if the code to log them in didn't work). But at the moment this isn't a problem and doing this at the class level will be faster than rebuilding the browser for each test, so let's keep this for now.

nkzarrabi replied:
django.test.TestCase does a full database flush at the start of each
new test. This means that you don't need to manually delete objects in
tearDown - the next test setUp will make sure the database is clean.

doctests and unittest.TestCase do no database preparations. At the
start of a test, the database will be in whatever state the previous
test left. This could mean some stray data which could cause testing
conflicts if you are not careful, so a cleanup might be good practice.

Yours,
Russ Magee %-)

https://groups.google.com/g/django-users/c/NNzptyVgzvg?pli=1

kerkeslager replied:
Yeah, I'm aware of the DB flush but I'm not sure browser settings would be flushed. It's not a big enough problem to worry about at the moment though.

nkzarrabi commented on core/test_integration.py:
Got rid of the useless "setUp(self)"  in class IntegrationTests

nkzarrabi reviewed (commented):
(env) PS C:\Users\zarra\Documents\GitHub\friendzone> python manage.py test core.test_integration                                                 
Found 3 test(s).
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
s
DevTools listening on ws://127.0.0.1:65232/devtools/browser/0320a3d9-4133-4e60-a49c-16f969882665
.[01/Feb/2024 14:27:38,611] - Broken pipe from ('127.0.0.1', 65244)
[01/Feb/2024 14:27:38,612] - Broken pipe from ('127.0.0.1', 65243)
[01/Feb/2024 14:27:38,615] - Broken pipe from ('127.0.0.1', 65254)
.
----------------------------------------------------------------------
Ran 2 tests in 19.013s

OK (skipped=1)

nkzarrabi reviewed (commented):
added user signup tests*

nkzarrabi commented:
@kerkeslager I added sign up tests yesterday but there were a couple of things wrong. I had duplicate code and also I made some comments addressing some concerns I only thought of reviewing my work today. 
Namely, sign up tests right now access the signup link directly and I'm not sure if I'm accessing it right.

kerkeslager commented:
I merged a different PR containing basically this code, so closing this.
